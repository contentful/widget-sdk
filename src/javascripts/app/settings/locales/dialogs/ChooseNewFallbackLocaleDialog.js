import React from 'react';
import PropTypes from 'prop-types';
import { Select, Option, ModalConfirm } from '@contentful/forma-36-react-components';

const LocaleType = PropTypes.shape({
  code: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
});

export default class ChooseNewFallbackLocaleDialog extends React.Component {
  static propTypes = {
    isShown: PropTypes.bool.isRequired,
    locale: LocaleType.isRequired,
    availableLocales: PropTypes.arrayOf(LocaleType.isRequired).isRequired,
    dependantLocales: PropTypes.arrayOf(LocaleType.isRequired).isRequired,
    onCancel: PropTypes.func.isRequired,
    onConfirm: PropTypes.func.isRequired,
  };

  state = {
    newFallbackCode: '',
  };

  componentDidUpdate(nextProps) {
    if (nextProps.isShown === false) {
      this.setState({ newFallbackCode: '' });
    }
  }

  onConfirm = () => {
    this.props.onConfirm(this.state.newFallbackCode);
  };

  prepareDependantLocaleNames = (dependantLocales) => {
    if (dependantLocales.length > 4) {
      const rest = ` and ${dependantLocales.length - 3} other locales`;
      return (
        dependantLocales
          .slice(0, 3)
          .map((item) => item.name)
          .join(', ') + rest
      );
    } else {
      return dependantLocales.map((item) => item.name).join(', ');
    }
  };

  render() {
    const { locale, dependantLocales, availableLocales } = this.props;
    return (
      <ModalConfirm
        isShown={this.props.isShown}
        size="large"
        title="You’re deleting a locale used as fallback"
        onCancel={this.props.onCancel}
        onConfirm={this.onConfirm}
        confirmTestId="choose-locale-confirm"
        cancelTestId="choose-locale-cancel"
        confirmLabel={`OK, delete ${locale.name} and update affected locales`}
        cancelLabel="Don’t delete">
        <p>
          <strong>{locale.name}</strong> is used as fallback for{' '}
          <code>{this.prepareDependantLocaleNames(dependantLocales)}</code>. After deleting it, the
          affected locales will fallback to:
        </p>
        <Select
          id="newFallbackLocale"
          name="newFallbackLocale"
          value={this.state.newFallbackCode}
          testId="choose-fallback-locale-select"
          onChange={(e) => {
            this.setState({ newFallbackCode: e.target.value });
          }}>
          <Option value="">None (no fallback)</Option>
          {availableLocales.map((locale) => (
            <Option key={locale.code} value={locale.code}>
              {locale.name}
            </Option>
          ))}
        </Select>
      </ModalConfirm>
    );
  }
}
