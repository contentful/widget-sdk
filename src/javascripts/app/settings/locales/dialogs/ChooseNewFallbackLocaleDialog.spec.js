import React from 'react';
import { shallow, mount } from 'enzyme';
import ChooseNewFallbackLocaleDialog from './ChooseNewFallbackLocaleDialog.es6';

describe('locales/components/ChooseNewFallbackLocaleDialog', () => {
  const selectors = {
    confirmChangeLocale: '[data-test-id="choose-locale-confirm"]',
    cancelChangeLocale: '[data-test-id="choose-locale-cancel"]',
    chooseLocaleSelect: '[data-test-id="choose-fallback-locale-select"]'
  };

  const renderComponent = props => (
    <ChooseNewFallbackLocaleDialog
      isShown
      onConfirm={() => {}}
      onCancel={() => {}}
      locale={{
        name: 'German',
        code: 'de'
      }}
      dependantLocales={[
        {
          name: 'Polish',
          code: 'pl'
        },
        {
          name: 'French',
          code: 'fr'
        }
      ]}
      availableLocales={[
        {
          name: 'Russian',
          code: 'ru'
        },
        {
          name: 'English',
          code: 'en'
        }
      ]}
      {...props}
    />
  );

  it('should match snapshot', () => {
    const wrapper = shallow(renderComponent());
    expect(wrapper).toMatchSnapshot();
  });

  it('it is possible to invoke cancel by clicking on two buttons', () => {
    const stubs = {
      onCancel: jest.fn()
    };
    const wrapper = mount(
      renderComponent({
        onCancel: stubs.onCancel
      })
    );

    wrapper.find(selectors.cancelChangeLocale).simulate('click');
    expect(stubs.onCancel).toHaveBeenCalledTimes(1);
  });

  it('click on submit sends current selected code to the callback', () => {
    const stubs = {
      onConfirm: jest.fn()
    };
    const wrapper = mount(
      renderComponent({
        onConfirm: stubs.onConfirm
      })
    );
    // click with 'none' selected
    wrapper.find(selectors.confirmChangeLocale).simulate('click');
    expect(stubs.onConfirm).toHaveBeenCalledWith('');
    // select 'en' and click again
    wrapper.find(selectors.chooseLocaleSelect).simulate('change', { target: { value: 'en' } });
    wrapper.find(selectors.confirmChangeLocale).simulate('click');
    expect(stubs.onConfirm).toHaveBeenCalledWith('en');
    expect(stubs.onConfirm).toHaveBeenCalledTimes(2);
  });
});
