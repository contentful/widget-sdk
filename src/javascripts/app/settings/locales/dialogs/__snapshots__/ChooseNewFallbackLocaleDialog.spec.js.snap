// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`locales/components/ChooseNewFallbackLocaleDialog should match snapshot 1`] = `
<ModalConfirm
  cancelLabel="Don’t delete"
  cancelTestId="choose-locale-cancel"
  confirmLabel="OK, delete German and update affected locales"
  confirmTestId="choose-locale-confirm"
  intent="positive"
  isConfirmDisabled={false}
  isConfirmLoading={false}
  isShown={true}
  onCancel={[Function]}
  onConfirm={[Function]}
  shouldCloseOnEscapePress={true}
  shouldCloseOnOverlayClick={true}
  size="large"
  testId="cf-ui-modal-confirm"
  title="You’re deleting a locale used as fallback"
>
  <p>
    <strong>
      German
    </strong>
     is used as fallback for
     
    <code>
      Polish, French
    </code>
    . After deleting it, the affected locales will fallback to:
  </p>
  <Select
    hasError={false}
    id="newFallbackLocale"
    isDisabled={false}
    name="newFallbackLocale"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    required={false}
    testId="choose-fallback-locale-select"
    value=""
    width="full"
  >
    <Option
      value=""
    >
      None (no fallback)
    </Option>
    <Option
      key="ru"
      value="ru"
    >
      Russian
    </Option>
    <Option
      key="en"
      value="en"
    >
      English
    </Option>
  </Select>
</ModalConfirm>
`;
